tensorflow/core/common_runtime/dml/dml_command_recorder.cc

```diff
@@ -370,12 +375,9 @@ void DmlCommandRecorder::CloseAndExecute() {
 
   // Fail early if something horrifying happens
   if (FAILED(device_removed_reason)) {
-    status_ = errors::Unknown(
-        "Device was removed because of the following reason: ",
-        dml_util::StringifyDeviceRemovedReason(device_removed_reason),
-        ". This can happen when the GPU times out or when there's a problem in "
-        "the driver. You won't be able to use this DML device again in the "
-        "current process.");
+    // Signal to the listeners that the device has been removed, so they can
+    // handle it appropriately (e.g. release memory)
+    device_removed_event_->NotifyListeners();
     return;
```

Comment:
Introducing this kind of re-entrancy is often difficult to reason about, because you don't know what the state of the world is when this callback gets invoked. The command recorder is the one who decides when to invoke this, but how do the listeners know whether it's safe to do something (e.g. release all memory) without having to understand the implementation details of the command recorder?

Don't we already have a central place where we check for device removal anyway? (Presumably sometime around when we flush) Can we just respond to the device removal at that point?

Reply:
Pooled heaps need to be notified of device removals to release the memory.

The listeners know it's "safe" to do something like releasing memory because it knows the state of the device is now removed, so nothing should try to use that memory anymore or hold on to it. We're not attempting to rescue the device here, but merely fail in a way that allows other devices to keep going.